WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/contactsselect/ContactsSelectAdapter.kt

```diff
@@ -0,0 +1,180 @@
+                return NewGroupViewHolder(view)
+            }
+        }
+    }
+
+    class NewContactViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+        // Set QRcode
+        val qrIv: ImageView = view.findViewById(R.id.ivQR)
+        init {
+            qrIv.setImageResource(R.drawable.ic_qrcode)
+            Log.i(LOGTAG, "ITEM ID: $itemId (NewContactsVH)")
+        }
+        companion object {
+            fun from(parent: ViewGroup): NewContactViewHolder {
+                val layoutInflater = LayoutInflater.from(parent.context)
+                val view = layoutInflater.inflate(R.layout.item_new_contact, parent, false)
+                return NewContactViewHolder(view)
+            }
+        }
+    }
+
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+//        Log.i(LOGTAG, "Position $position - onBindVH")
+        when (holder) {
+            is ContactsViewHolders -> {
+                val newPosition = position - ADDEDED_HEADERS
+                val item = getItem(newPosition)
+                if (item != null) {
+                    tracker?.let {
+                        holder.bind(getItem(newPosition), it.isSelected(newPosition.toLong()))
+                    }
+//                    holder.bind(item)
+                    Log.i(LOGTAG, "$item Position: $newPosition")
+                }
+
+            }
+        }
+    }
+
+    class DiffCallback : DiffUtil.ItemCallback<Contact>(){
+        override fun areItemsTheSame(oldItem: Contact, newItem: Contact): Boolean {
+            return oldItem == newItem
+        }
+
+        override fun areContentsTheSame(oldItem: Contact, newItem: Contact): Boolean {
+            return oldItem.uid == newItem.uid
+        }
+
+    }
+
+    override fun getItemCount(): Int {
+        return super.getItemCount() + ADDEDED_HEADERS
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        return if (position == 0) {
+            ITEM_VIEW_TYPE_NEW_GROUP
+        } else if (position == 1) {
+            ITEM_VIEW_TYPE_NEW_CONTACT
+        } else {
+            ITEM_VIEW_TYPE_CONTACTS
+        }
+    }
+
+    override fun getItemId(position: Int): Long {
+        return (position - ADDEDED_HEADERS).toLong()
+    }
+}
+
+class ContactsKeyProvider(private val recyclerView: RecyclerView) :
+    ItemKeyProvider<Long>(SCOPE_MAPPED) {
+    override fun getKey(position: Int): Long? {
+        val theItemId = recyclerView.adapter?.getItemId(position)
+        Log.i(LOGTAG, "GetKey: $theItemId")
+        return theItemId
+    }
+
+    override fun getPosition(key: Long): Int {
+        val viewHolder = recyclerView.findViewHolderForItemId(key)
+        return viewHolder?.layoutPosition ?: RecyclerView.NO_POSITION
+    }
+}
+
+class ContactsDetailsLookup(private val recyclerView: RecyclerView) :
+    ItemDetailsLookup<Long>() {
+    override fun getItemDetails(event: MotionEvent): ItemDetails<Long>? {
```

Comment:
```
    override fun getItemDetails(event: MotionEvent): ItemDetails<Long>? {
        val view: View? = recyclerView.findChildViewUnder(event.x, event.y)
        view?.let {
            val viewTrack = recyclerView.getChildViewHolder(it)
            Log.i(LOGTAG, "View: ${viewTrack?.itemId}")

            if (viewTrack.itemId >= 0) {
                return (recyclerView.getChildViewHolder(view) as ContactsSelectAdapter.ContactsViewHolders)
                    .getItemDetails()
            }
        }
        return null
    }

I think something like this might be more readable/easier to follow
```